File: Algorithm/JumpGameVI.cpp
// // Time complexity: O(n)
// // Space complexity: O(n+k) -> O(k)
// class Solution {
// public:
//     int maxResult(vector<int>& nums, int k) {
//         const int n = nums.size();
//         vector<int> dp(n);
//         deque<int> q{{0}};
//         dp[0] = nums[0];
//         for (int i = 1; i < n; ++i) {
//             dp[i] = nums[i] + dp[q.front()];
//             while (!q.empty() && dp[i] >= dp[q.back()]) q.pop_back();
//             while (!q.empty() && i - q.front() >= k) q.pop_front();
//             q.push_back(i);
//         }
//         return dp[n - 1];
//     }
// };

// Space complexity: O(k)
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        const int n = nums.size();
        deque<pair<int, int>> q{{nums[0], 0}};
        for (int i = 1; i < n; ++i) {
            const int cur = nums[i] + q.front().first;
            while (!q.empty() && cur >= q.back().first)
                q.pop_back();
            while (!q.empty() && i - q.front().second >= k)
                q.pop_front();
            q.emplace_back(cur, i);
        }
        for (const auto& [v, i] : q)
            if (i == n - 1) return v;
        return 0;
    }
};

File: Algorithm/JumpGameVI.java
// // Solution: DP + Monotonic Queue
// // dp[i] = nums[i] + max(dp[j]) i - k <= j < i
// // Brute force time complexity: O(n*k) => TLE
// // This problem can be reduced to find the maximum for a sliding window that can be solved by monotonic queue.
// // Time complexity: O(n)
// // Space complexity: O(n+k) -> O(k)
// class Solution {
//     public int maxResult(int[] nums, int k) {
//         final int n = nums.length;
//         Deque<Integer> q = new LinkedList<Integer>(Arrays.asList(0));
//         int[] dp = new int[n];
//         dp[0] = nums[0];
//         for (int i = 1; i < n; ++i) {
//             dp[i] = nums[i] + dp[q.peekFirst()];
//             while (!q.isEmpty() && dp[i] >= dp[q.peekLast()]) q.pollLast();
//             while (!q.isEmpty() && i - q.peekFirst() >= k) q.pollFirst();
//             q.offerLast(i);
//         }
//         return dp[n - 1];
//     }
// }

class Solution {
    public int maxResult(int[] nums, int k) {
        final int n = nums.length;
        Deque<int[]> q = new LinkedList<int[]>();
        q.offer(new int[]{nums[0], 0});
        for (int i = 1; i < n; ++i) {
            final int cur = nums[i] + q.peekFirst()[0];
            while (!q.isEmpty() && cur >= q.peekLast()[0]) 
                q.pollLast();
            while (!q.isEmpty() && i - q.peekFirst()[1] >= k) 
                q.pollFirst();
            q.offerLast(new int[]{cur, i});
        }
        for (int[] vi : q) 
            if (vi[1] == n - 1) return vi[0];
        return 0;
    }
}
